File: snippets/java/src/main/java/develop/ServiceCommunication.java
package develop;

import dev.restate.common.InvocationOptions;
import dev.restate.common.Request;
import dev.restate.common.Target;
import dev.restate.sdk.Restate;
import dev.restate.serde.TypeTag;
import java.time.Duration;

public class ServiceCommunication {

  private void requestResponseService() {
    String request = "";
    String workflowId = "";
    String objectKey = "";

    // <start_request_response>
    // --- Simple client: awaits inline and returns the result directly ---
    // To call a Service:
    String svcResponse = Restate.service(MyService.class).myHandler(request);
    // To call a Virtual Object:
    String objResponse = Restate.virtualObject(MyObject.class, objectKey).myHandler(request);
    // To call a Workflow:
    String wfResponse = Restate.workflow(MyWorkflow.class, workflowId).run(request);

    // --- Handle-based client: returns a DurableFuture to await explicitly ---
    // Use it for invocation options (e.g. an idempotency key), timeouts, or concurrency.
    String svcResult =
        Restate.serviceHandle(MyService.class).call(MyService::myHandler, request).await();
    String objResult =
        Restate.virtualObjectHandle(MyObject.class, objectKey)
            .call(MyObject::myHandler, request)
            .await();
    String wfResult =
        Restate.workflowHandle(MyWorkflow.class, workflowId).call(MyWorkflow::run, request).await();
    // <end_request_response>
  }

  private void genericCall() {
    String request = "";

    // <start_request_response_generic>
    Target target = Target.service("MyService", "myHandler"); // or virtualObject or workflow
    String response =
        Restate.call(
                Request.of(target, TypeTag.of(String.class), TypeTag.of(String.class), request))
            .await();
    // <end_request_response_generic>
  }

  private void genericSend() {
    String request = "";

    // <start_one_way_generic>
    Target target = Target.service("MyService", "myHandler"); // or virtualObject or workflow
    Restate.send(Request.of(target, TypeTag.of(String.class), TypeTag.of(String.class), request));
    // <end_one_way_generic>
  }

  private void genericDelayedSend() {
    String request = "";

    // <start_delayed_generic>
    Target target = Target.service("MyService", "myHandler"); // or virtualObject or workflow
    Restate.send(
        Request.of(target, TypeTag.of(String.class), TypeTag.of(String.class), request),
        Duration.ofDays(5));
    // <end_delayed_generic>
  }

  private void oneWay() {
    String request = "";

    // <start_one_way>
    Restate.serviceHandle(MyService.class).send(MyService::myHandler, request);
    // <end_one_way>
  }

  private void idempotencyKey() {
    String request = "";

    // <start_idempotency_key>
    // For a request-response call
    Restate.serviceHandle(MyService.class)
        .call(MyService::myHandler, request, InvocationOptions.idempotencyKey("abc123"));
    // For a message
    Restate.serviceHandle(MyService.class)
        .send(MyService::myHandler, request, InvocationOptions.idempotencyKey("abc123"));
    // <end_idempotency_key>
  }

  private void attach() {
    String request = "";

    // <start_attach>
    var handle =
        Restate.serviceHandle(MyService.class)
            .send(MyService::myHandler, request, InvocationOptions.idempotencyKey("abc123"));
    var response = handle.attach().await();
    // <end_attach>
  }

  private void cancel() {
    String request = "";

    // <start_cancel>
    var handle = Restate.serviceHandle(MyService.class).send(MyService::myHandler, request);
    handle.cancel();
    // <end_cancel>
  }

  private void delayedCall() {
    String request = "";

    // <start_delayed>
    Restate.serviceHandle(MyService.class).send(MyService::myHandler, request, Duration.ofDays(5));
    // <end_delayed>
  }

  private void orderingGuarantees() {
    String objectKey = "";
    // <start_ordering>
    Restate.virtualObjectHandle(MyObject.class, objectKey).send(MyObject::myHandler, "I'm call A");
    Restate.virtualObjectHandle(MyObject.class, objectKey).send(MyObject::myHandler, "I'm call B");
    // <end_ordering>
  }
}

File: snippets/kotlin/src/main/kotlin/develop/ServiceCommunication.kt
package develop

import dev.restate.common.Request
import dev.restate.common.Target
import dev.restate.sdk.kotlin.*
import dev.restate.serde.kotlinx.*
import kotlin.time.Duration.Companion.days

class ServiceCommunication {
  suspend fun requestResponseService() {
    val request = ""
    val objectKey = ""
    val workflowId = ""

    // <start_request_response>
    // --- Simple client: awaits inline and returns the result directly ---
    // To call a Service:
    val svcResponse = service<MyService>().myHandler(request)
    // To call a Virtual Object:
    val objResponse = virtualObject<MyObject>(objectKey).myHandler(request)
    // To call a Workflow:
    val wfResponse = workflow<MyWorkflow>(workflowId).run(request)

    // --- Handle-based client: returns a DurableFuture to await explicitly ---
    // Use it for invocation options (e.g. an idempotency key), timeouts, or concurrency.
    val svcResult = toService<MyService>().request { myHandler(request) }.call().await()
    val objResult =
        toVirtualObject<MyObject>(objectKey).request { myHandler(request) }.call().await()
    val wfResult = toWorkflow<MyWorkflow>(workflowId).request { run(request) }.call().await()
    // <end_request_response>
  }

  suspend fun oneWay() {
    val request = ""

    // <start_one_way>
    toService<MyService>().request { myHandler(request) }.send()
    // <end_one_way>
  }

  suspend fun idempotencyKey() {
    val request = ""

    // <start_idempotency_key>
    // For a regular call
    toService<MyService>()
        .request { myHandler(request) }
        .options { idempotencyKey = "abc123" }
        .call()
    // For a one way call
    toService<MyService>()
        .request { myHandler(request) }
        .options { idempotencyKey = "abc123" }
        .send()
    // <end_idempotency_key>
  }

  suspend fun attach() {
    val request = ""

    // <start_attach>
    val handle =
        toService<MyService>()
            .request { myHandler(request) }
            .options { idempotencyKey = "abc123" }
            .send()
    val response = handle.attach().await()
    // <end_attach>
  }

  suspend fun cancel() {
    val request = ""

    // <start_cancel>
    val handle = toService<MyService>().request { myHandler(request) }.send()
    handle.cancel()
    // <end_cancel>
  }

  suspend fun scope() {
    val request = ""
    val objectKey = ""
    val workflowId = ""

    // <start_scope>
    // Route a call into a named scope
    val svcResponse = scope("tenant-123").service<MyService>().myHandler(request)

    // Add a limit key for hierarchical concurrency limits within the scope
    val wfResponse =
        scope("tenant-123")
            .toWorkflow<MyWorkflow>(workflowId)
            .request { run(request) }
            .options { limitKey = "premium/user42" }
            .call()
            .await()

    // Scoped Virtual Object calls need
    // RESTATE_EXPERIMENTAL_ENABLE_SCOPED_VIRTUAL_OBJECTS=true on the server
    val objResponse = scope("tenant-123").virtualObject<MyObject>(objectKey).myHandler(request)
    // <end_scope>

    // <start_scope_request>
    // The scope and limit key the invocation was submitted with
    val invocationScope = request().scope
    val invocationLimitKey = request().limitKey
    // <end_scope_request>
  }

  suspend fun generic() {
    val request = ""

    // <start_request_response_generic>
    val target = Target.service("MyService", "myHandler")
    val response =
        prepareRequest(Request.of(target, typeTag<String>(), typeTag<String>(), request))
            .call()
            .await()
    // <end_request_response_generic>
  }

  suspend fun genericSend() {
    val request = ""

    // <start_one_way_generic>
    val target = Target.service("MyService", "myHandler")
    prepareRequest(Request.of(target, typeTag<String>(), typeTag<String>(), request)).send()
    // <end_one_way_generic>
  }

  suspend fun genericDelayed() {
    val request = ""

    // <start_delayed_generic>
    val target = Target.service("MyService", "myHandler")
    prepareRequest(Request.of(target, typeTag<String>(), typeTag<String>(), request))
        .send(delay = 5.days)
    // <end_delayed_generic>

  }

  suspend fun delayedCall() {
    val request = ""

    // <start_delayed>
    toService<MyService>().request { myHandler(request) }.send(5.days)
    // <end_delayed>
  }
}
